forge-game/src/main/java/forge/game/ability/AbilityUtils.java

```diff
@@ -1352,6 +1354,28 @@ public static void resolve(final SpellAbility sa) {
         AbilityUtils.resolveApiAbility(sa, game);
     }
 
+    private static void resolvePreAbilities(final SpellAbility sa, final Game game) {
+        Player controller = sa.getActivatingPlayer();
+        Card source = sa.getHostCard();
+        // do blessing there before condition checks
+        if (sa.isSpell() && !source.isPermanent() && source.hasKeyword(Keyword.ASCEND)) {
+            if (controller.getZone(ZoneType.Battlefield).size() >= 10) {
+                controller.setBlessing(true);
+            }
+        }
+
+        if (sa.isSpell() && !(sa instanceof SpellPermanent) && sa.getHostCard().hasKeyword(Keyword.GIFT) && sa.isOptionalCostPaid(OptionalCost.PromiseGift)) {
```

Comment:
I don't know if `!source.isPermanent()` wouldn't be better like above with Ascend

Reply:
Is source.isPermanent right? Does that calculate properly for things like Adventure?

Reply:
At this place, shortly before the Spell is going to resolve, 
the Spell Host has changed into the Adventure Spell Half (so it is not a Permanent, and it shouldn't have had Gift anyway)